packages/core/test/Form.test.jsx

```diff
@@ -3716,13 +3719,192 @@ describeRepeated('Form common', (createFormComponent) => {
+    });
+    const element = node.querySelector('input[type=text]');
+    fireEvent.change(element, {
+      target: { value: 'short' },
+    });
+    fireEvent.blur(element);
+
+    sinon.assert.calledWithMatch(onBlur.lastCall, 'root', 'short');
+    sinon.assert.calledWithMatch(
+      onChange.lastCall,
+      {
+        formData: 'short',
+        errorSchema: {
+          __errors: ['must NOT have fewer than 8 characters'],
+        },
+      },
+      'root',
+    );
+  });
+});
 
+describe('omitExtraData and live omit onBlur', () => {
+  let sandbox;
   beforeEach(() => {
     sandbox = createSandbox();
   });
+  afterEach(() => {
+    sandbox.restore();
+  });
+  const schema = {
+    type: 'object',
+    properties: {
+      foo: { type: 'string' },
+      bar: { type: 'string' },
+    },
+  };
+  const formData = { foo: 'foo', bar: 'bar' };
+
+  it('does not occur during onChange, empty data not removed', () => {
+    const onBlur = sinon.spy();
+    const { node, onChange } = createFormComponent({
+      schema,
+      formData,
+      onBlur,
+      omitExtraData: true,
+      liveOmit: 'onBlur',
+    });
+
+    fireEvent.change(node.querySelector('#root_foo'), {
+      target: { value: '' },
+    });
+
+    sinon.assert.notCalled(onBlur);
+    sinon.assert.calledWithMatch(
+      onChange.lastCall,
+      {
+        formData: { ...formData, foo: undefined },
+      },
+      'root_foo',
+    );
+  });
+
+  it('occurs during onBlur, onChange not last called with empty data removed due to no state update', () => {
+    const onBlur = sinon.spy();
+    const { node, onChange } = createFormComponent({
+      schema,
+      formData,
+      onBlur,
+      omitExtraData: true,
+      liveOmit: 'onBlur',
+    });
+    const element = node.querySelector('input[type=text]');
+    fireEvent.change(node.querySelector('#root_foo'), {
+      target: { value: '' },
+    });
+    fireEvent.change(node.querySelector('#root_foo'), {
+      target: { value: 'foo' },
+    });
+    fireEvent.blur(element);
+
+    sinon.assert.calledWithMatch(onBlur.lastCall, 'root', 'foo');
+    sinon.assert.calledWithMatch(
+      onChange.lastCall,
+      {
+        formData,
+      },
+      'root',
+    );
+  });
+
+  it('occurs during onBlur, onChange last called with empty data removed due to a state update', () => {
+    const onBlur = sinon.spy();
+    const { node, onChange } = createFormComponent({
+      schema,
+      formData,
+      onBlur,
+      omitExtraData: true,
+      liveOmit: 'onBlur',
+    });
+    const element = node.querySelector('input[type=text]');
+    fireEvent.change(node.querySelector('#root_foo'), {
+      target: { value: '' },
+    });
+    fireEvent.blur(element);
+
+    sinon.assert.calledWithMatch(onBlur.lastCall, 'root', '');
+    sinon.assert.calledWithMatch(
+      onChange.lastCall,
+      {
+        formData: { bar: 'bar' },
+      },
+      'root',
+    );
+  });
+});
 
+describe('Form omitExtraData and liveOmit', () => {
```

Comment:
Everything else from here down is just a reorganization. Hiding whitespace will help reduce a lot of changed lines